stdlib.sh

```diff
@@ -61,6 +61,113 @@ puts_verbose() {
   fi
 }
 
+# Key Value Store Utility
+# -------------------
+
+# This is a utility module that allows you to create a file that serves
+# as a key-value store that can be persisted to disk.
+#
+# It works around using lines formatted like:
+# 
+# key=value
+#
+# and using grep when searching for keys to find the last entry. This 
+# allows an existing key to be overwritten.
+#
+# This is inspired by the "world's simplest database" in Ch 3 of
+# Designing Data-Intensive Applications https://dataintensive.net/
+
+# Ensure that the file that we would like to persist this store to
+# exists. Also handles creating the containing directory if it doesn't
+# exist yet. 
+#
+# Will not overwrite the store if it already exists
+# Usage: kv_create file
+kv_create() {
+  local f=$1
+  mkdir -p "$(dirname "$f")"
+  touch "$f"
+}
+
+# Delete all keys and values from the store
+# Usage: kv_clear file
+kv_clear() {
+  local f=$1
+  echo "" > "$f"
+}
+
+# Set value for a key
+# Usage: kv_set file key value
+kv_set() {
+  if [[ $# -eq 3 ]]; then
+    local f=$1
+    if [[ -f $f ]]; then
+      echo "$2=$3" >> "$f"
+    fi
+  fi
+}
+
+# Retrieve value for a key. Will return an empty string if the key is
+# not present
+# Usage: kv_get file key
+kv_get() {
+  if [[ $# -eq 2 ]]; then
+    local f=$1
+    if [[ -f $f ]]; then
+      grep "^$2=" "$f" | sed -e "s/^$2=//" | tail -n 1
+    fi
+  fi
+}
+
+# Retrive the value for a key, but wrap it in quotes if the value contains spaces
+# This is expected to be mostly used internally
+# Usage: kv_get_escaped file key
+kv_get_escaped() {
+  local value
+  value=$(kv_get "$1" "$2")
+  if [[ $value =~ [[:space:]]+ ]]; then
+    echo "\"$value\""
+  else
+    echo "$value"
+  fi
+}
+
+# List out each of the keys in the store
+# Usage: kv_keys file
+kv_keys() {
+  local f=$1
+  local keys=()
+
+  if [[ -f $f ]]; then
+    # Iterate over each line, splitting on the '=' character
+    #
+    # The || [[ -n "$key" ]] statement addresses an issue with reading the last line
+    # of a file when there is no newline at the end. This will not happen if the file
+    # is created with this module, but can happen if it is written by hand.
+    # See: https://stackoverflow.com/questions/12916352/shell-script-read-missing-last-line
+    while IFS="=" read -r key value || [[ -n "$key" ]]; do
+      # if there are any empty lines in the store, skip them
+      if [[ -n $key ]]; then
+        keys+=("$key")
+      fi
+    done < "$f"
+
+    echo "${keys[@]}" | tr ' ' '\n' | sort -u
+  fi
+}
+
+# Log out each of the keys and values in the store, one per line
+# Usage: kv_list file
+kv_list() {
+  local f=$1
+
+  kv_keys "$f" | tr ' ' '\n' | while read -r key; do
+    if [[ -n $key ]]; then
+      echo "$key=$(kv_get_escaped "$f" "$key")"
```

Comment:
you could maybe also use `printf %q "$value"` instead of this escaping stuff

Reply:
I tried this, but could not get it to unescape with `printf %b`